proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,1330 @@
+  because the ambient QEnv in ``M`` includes both ``L.Internal.z`` and ``Internal.z``,
+  both referring to the same original name.
+
+  There is thus the small potential for abstraction leakage.
+
+  It would be possible to introduce a warning to detect this scenario.
+
+Related Work
+------------
+
+There is much prior art. The list below is shamelessly cribbed from `#205`_.
+
+* Proposal `#205`_. That proposal essentially tweaks the ``qualified`` feature to
+  become more flexible and exportable. It has proved hard to digest (from the commentary),
+  though, and solves fewer problems than this proposal. On the other hand, it is likely
+  easier to implement.
+
+  This proposal is inspired by some of the topics that came up in the conversation
+  for `#205`_, and I'm grateful for @deepfire's efforts on that proposal.
+
+* 2005 Coutts, ``as`` in export lists: `<https://mail.haskell.org/pipermail/libraries/2005-March/003390.html>`_ . Salient points:
+  letting modules export other modules' contents qualified with the module name`
+
+* 2006 Wallace, explicit namespaces for module names: `<https://ghc.haskell.org/trac/ghc/wiki/Commentary/Packages/PackageNamespacesProposal>`_ . Salient points:
+  The declaration import namespace brings into availability the subset of the hierarchy of module names rooted in the package "foo-1.3", at the position ``Data.Foo``
+
+* 2013 de Castro Lopo, qualified exports: `<https://wiki.haskell.org/GHC/QualifiedModuleExport>`_
+  ``qualified module T`` in export list and is essentially a subset of this proposal.
+
+* A worthwhile counter-proposal can be found at `#295`_. That proposal re-casts modules as
+  entities proper; this one, in contrast, continues the historical treatment of modules
+  simply as qualifications to identifiers. In my view, `#295`_ needs to do too much work
+  to keep backward compatibility due to its more fundamental approach. Given a fresh start,
+  I would likely prefer something like `#295`_ than what I have written here, but we're not
+  making a fresh start.
+
+Alternatives
+------------
+
+Beyond the `Related Work`_, there is wiggle room within this proposal for alternatives.
+
+1. Disallow local modules to be mutually recursive. The current proposal says
+   that the local module system affects scoping only. However, we could instead declare
+   that a mutual-dependency strongly-connected component (SCC) cannot include definitions
+   in more than one module. This would disable mutual recursion between modules, but open
+   the possibility of using local modules to explicitly stage compilation. This might
+   allow, for example, the definition of a function and its usage in a Template Haskell
+   splice in the same file (as long as they were in different local modules).
+
+   I prefer the proposal as-is in this regard: modules should be about abstraction and
+   naming, not about compilation dependencies. Compilation dependencies should be handled
```

Comment:
Abstraction, naming, and **scoping**, which is why intermodule mutual recursion seems so bizarre to me.

Reply:
I don't see the relationship between scoping and mutual recursion. If an entity is exported from a local module (which is the default), it is in scope throughout the enclosing context of the local module. This would allow for mutual recursion. I'm worried I'm missing something here.

Reply:
I guess my mental model for how local modules would work is that they should resemble a `let` block, e.g.

```
module Top where
  module A where
    x = 5
    y = True

  module B where
    z = 42
```

should behave similarly to

```
module Top where

(x, y) = 
  let 
    x = 5
    y = True
  in (x, y)

z =
  let
    z = 42
  in z
```

(let's set aside for the purpose of this thread whether the top-level x,y,z are qualified or not)

If local modules are morally desugarable to `let`-blocks as I've sketched, we certainly wouldn't allow mutual recursion across them.

But perhaps it's I who am missing something. You mentioned above that (local?) modules are not entities, merely namespaces. I'm not sure I agree with that view, but allowing inter-module mutual recursion certainly makes more sense from that standpoint.

Reply:
I think those `let`-blocks *can* be mutually recursive:

```hs
(x, y) =
  let
    x = 5
    y = z : [True]
  in (x, y)

z =
  let
    z = head (tail y)
  in z
```

This works today.

Reply:
Hrm, that's a bit cheeky, but you're right it does make sense. I guess I withdraw my concern.